File: ui-mobile/src/features/devices/mappers/devices.mapper.ts
import { DeviceResponse, RegisterDeviceInput } from "@/features/devices/types/devices.types";

export const devicesMapper = {
  toRequest: (raw: RegisterDeviceInput): RegisterDeviceInput => ({
    platformCode: raw.platformCode,
    pushToken: raw.pushToken ?? null,
    deviceName: raw.deviceName ?? null,
    appVersion: raw.appVersion ?? null,
    osVersion: raw.osVersion ?? null
  }),
  fromResponse: (raw: DeviceResponse): DeviceResponse => ({ ...raw })
};

File: service/truespend.api/Mappers/DevicesMapper.cs
using TrueSpend.Api.ViewModels.Devices;
using DomainRegister = TrueSpend.Domain.Models.Devices.RegisterDeviceRequest;
using DomainUpdate = TrueSpend.Domain.Models.Devices.UpdateDeviceRequest;
using DomainDevice = TrueSpend.Domain.Models.Devices.DeviceResponse;

namespace TrueSpend.Api.Mappers;

public interface IDevicesMapper
{
    DomainRegister ToDomain(RegisterDeviceRequestVm request);
    DomainUpdate ToDomain(UpdateDeviceRequestVm request);
    DeviceResponseVm ToResponse(DomainDevice domain);
}

public sealed class DevicesMapper : IDevicesMapper
{
    public DomainRegister ToDomain(RegisterDeviceRequestVm request) =>
        new(request.PlatformCode, request.PushToken, request.DeviceName, request.AppVersion, request.OsVersion, request.Locale, request.Timezone, request.InstallationId);

    public DomainUpdate ToDomain(UpdateDeviceRequestVm request) =>
        new(request.PushToken, request.DeviceName, request.AppVersion, request.OsVersion, request.Locale, request.Timezone);

    public DeviceResponseVm ToResponse(DomainDevice domain) => new(domain.DeviceId, domain.Registered);
}
